artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPConnectionContext.java

```diff
@@ -183,6 +191,17 @@ public void flush() {
    }
 
    public void close(ErrorCondition errorCondition) {
+      synchronized (schedulingLock) {
+         isSchedulingCancelled = true;
+
+         if (scheduledPool != null && scheduledPool instanceof ThreadPoolExecutor &&
+            scheduledFuture != null && scheduledFuture instanceof Runnable) {
+            if (!((ThreadPoolExecutor) scheduledPool).remove((Runnable) scheduledFuture)) {
```

Comment:
The other issue is perf-wise: if the work queue used on the `ScheduledExecutorService` is a `LinkedBlockingQueue` it requires a full lock on the queue to perform the remove, that's bad from a scalability point of view, given that the `ScheduledExecutorService` is shared between several components, see https://github.com/jdk-mirror/openjdk8/blob/master/java/util/concurrent/LinkedBlockingQueue.java#L548

Reply:
I see your point about performance but I think that sooner or later that remove will be done.
I also think it's important free all resource related an object before finish close for two matters:
1) free resource for the broker
2) avoid unpredictable behavior

Reply:
Yep, but the resources that can actually lead to OOM are not the Runnable, but the context that the Runnable(s) contains: If such context is referenced in an explicit, but `WeakReference` it would prevent it to be collected. Then `Runnable` will be then removed on the next (supposed) fire, without affecting scalability.

Reply:
Currently scheduledPool is used by:
1) AMQPConnectionContext more time but one task at once for instance
2) ActiveMQServerImpl one time and one task
3) SharedNothingBackupQuorum one time and one task
4) MessagePullHandler each time AMQConsumer.handleDeliver is called if prefetchSize = 0

ScheduledPool is mostly used by AMQPConnectionContext (one for each amqp connection) and remove is called only on connection close, so the effort for the remove should not impact on the entire system scalability.
Moreover GC should manage that unused objects until they are not removed from the scheduledPool.
Finally WeakReference require deeper source change for ScheduleRunnable and TickerRunnable to avoid unpredictable behavior when GC delete AMQPConnectionContext before their execution.

My concern is that to introduce unpredictable behavior to avoid paying a small computational cost.

Reply:
I won't guess which effects a remove would have globally: I just think is not a good idea to do it if not necessary, given that the scheduled pool usage span outside the context of this PR change.

> Finally WeakReference require deeper source change for ScheduleRunnable and TickerRunnable to avoid unpredictable behavior when GC delete AMQPConnectionContext before their execution.

Totally agree, but it would allow a more robust behaviour if a cancel won't get rid of the task instance. Anyway this last one (a missing cancel) is an additional issue and I agree it doesn't need to be addressed here.

My whole point is that having such casting cascade just to use the remove API makes me believe that worth investigating if this same PR could be addressed in a cleaner way (ie less hacky, not with less code changes) and evaluate if it is more portable/effective too.

Feel free to try addressing in another way and provide your branch to @k-wall to try it before merging it too :+1: